core/src/main/java/org/pentaho/di/connections/vfs/DefaultVFSConnectionFileNameTransformer.java

```diff
@@ -0,0 +1,192 @@
+import org.pentaho.di.connections.vfs.provider.ConnectionFileNameUtils;
+import org.pentaho.di.core.exception.KettleException;
+
+import java.util.Objects;
+
+import static org.apache.commons.vfs2.FileName.SEPARATOR;
+
+public class DefaultVFSConnectionFileNameTransformer<T extends VFSConnectionDetails>
+  implements VFSConnectionFileNameTransformer<T> {
+
+  @NonNull
+  private static final String SCHEME_SUFFIX = ":" + SEPARATOR + SEPARATOR;
+
+  @NonNull
+  private final VFSConnectionManagerHelper vfsConnectionManagerHelper;
+
+  @NonNull
+  private final ConnectionFileNameUtils connectionFileNameUtils;
+
+  public DefaultVFSConnectionFileNameTransformer() {
+    this( VFSConnectionManagerHelper.getInstance(), ConnectionFileNameUtils.getInstance() );
+  }
+
+  public DefaultVFSConnectionFileNameTransformer( @NonNull VFSConnectionManagerHelper vfsConnectionManagerHelper,
+                                                  @NonNull ConnectionFileNameUtils connectionFileNameUtils ) {
+    this.vfsConnectionManagerHelper = Objects.requireNonNull( vfsConnectionManagerHelper );
+    this.connectionFileNameUtils = Objects.requireNonNull( connectionFileNameUtils );
+  }
+
+  @NonNull
+  protected VFSConnectionManagerHelper getVfsConnectionManagerHelper() {
+    return vfsConnectionManagerHelper;
+  }
+
+  @NonNull
+  protected ConnectionFileNameUtils getConnectionFileNameUtils() {
+    return connectionFileNameUtils;
+  }
+
+  // region toProviderUri
+  @NonNull
+  @Override
+  public String toProviderUri( @NonNull ConnectionFileName pvfsFileName, @NonNull T details )
+    throws KettleException {
+
+    StringBuilder providerUriBuilder = new StringBuilder();
+
+    appendProviderUriConnectionRoot( providerUriBuilder, details );
+
+    // Examples:
+    //  providerUriBuilder: "hcp://domain.my:443/root/path" | "local:///C:/root/path" | "s3://"
+    //  getPath():          "/folder/sub-folder" | "/"
+
+    appendProviderUriRestPath( providerUriBuilder, pvfsFileName.getPath(), details );
+
+    // Examples: "hcp://domain.my:443/root/path/folder/sub-folder" | "s3://folder/sub-folder"
+
+    return providerUriBuilder.toString();
+  }
+
+  // IMPROVEMENT: The connection root provider URI is being repeatedly determined, but is actually constant as long as
+  // the connection details does not change its configuration. Ideally, some kind of caching would be used.
+  // Only the "rest path" differs each time. The problem being there is no way to know if connection details have
+  // changed. These don't even implement hashCode, atm. Could use the connection manager to store and manage this
+  // cache, by name, invalidating it whenever a connection is loaded/saved. Would still need to deal with in-memory
+  // connections used by test(..) somehow.
+  protected void appendProviderUriConnectionRoot( @NonNull StringBuilder providerUriBuilder, @NonNull T details )
+    throws KettleException {
+
+    appendProviderRootPrefix( providerUriBuilder, details );
+
```

Comment:
Regarding URI and URL, I don't think they can be used because of the special/reserved/encoded characters.
Regarding Path and Paths, I honestly am not familiar with the API and its implications/assumptions.

Some of the `ConnectionFileNameUtils` use `StringBuilder`, and using Path would require using strings. This might cause issues in `ConnectionFileNameParser`. For the transformer classes, we're creating the pattern now and could avoid using the `StringBuilder` pattern from the start. The `ConnectionFileNameUtils` are agnostic to any encoding issues, and simply look at, deal with separators.

The questions then being: how `Path(s)` would deal with the `pvfs` scheme and other Apache VFS registered schemes? Would it use the correct separator character? Does it have any encoding requirements?

Given time, I will look at this. In any case, any help on the above would be appreciated.